File: NeetCodeIO/110BalancedBinaryTree/BalancedBinaryTree.py
class Solution:
    def getHeight(self, root):
        if not root:
            return (0, True)
        leftHeight, leftBool = self.getHeight(root.left)
        rightHeight, rightBool = self.getHeight(root.right)
        return (1 + max(leftHeight, rightHeight), abs(leftHeight - rightHeight) <= 1 and leftBool and rightBool)

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        # DFS -> Time:O(n), Space:O(h)
        return self.getHeight(root)[1]

File: NeetCodeIO/110BalancedBinaryTree/BalancedBinaryTree.cpp
class Solution {
public:
    pair<int,bool> getHeight(TreeNode* root) {
        pair<int,bool> ans;
        if (!root) {
            ans.first = 0;
            ans.second = true;
            return ans;
        }
        pair<int,bool> left = getHeight(root->left);
        pair<int,bool> right = getHeight(root->right);
        ans.first = 1 + max(left.first, right.first);
        ans.second = abs(left.first - right.first) <= 1 && left.second && right.second;
        return ans;
    }
    bool isBalanced(TreeNode* root) {
        // DFS -> Time:O(n), Space:O(h)
        return getHeight(root).second;
    }
};
